### ISSUE TITLE
Multiple folders open in the sidebar

### ISSUE THREAD

**Issue by hmcts-jmoore** [marktext/marktext#2222]
### Describe your feature request
Multiple folders open in the sidebar at the same time.

### What problem does this feature solve? [optional]

This would help with working across multiple projects at the same time.  With only one folder openable at a time, I either have to keep switching folders or open a common root folder, which results in unwanted files and folders cluttering the sidebar.

**Comment by brainchild0** [comment]
Is it adequate to keep each folder open in a separate window?

**Comment by hmcts-jmoore** [comment]
> Is it adequate to keep each folder open in a separate window?

That works, but not as well for me.  I'm moving from using the Atom editor for markdown editing, which has both options, and I always end up using the single-window, multiple-folder version.  Personal preference - I don't know how many people would agree or disagree.

**Comment by brainchild0** [comment]
I installed Atom 1.48 on my Linux Mint system, and I find, when I open a folder after having already open one already, that the application creates a another window for the new folder, without affecting the open window, exactly as does Mark Text.

Would you give more details of how you are able to use Atom that you feel would express an appropriate enhancement for Mark Text?

**Comment by hmcts-jmoore** [comment]
In Atom, you have to "Add Project Folder" - just opening a new folder does open a new window.

I tend to work on a number of related codebases at the same time - all in the same editor instance to keep related stuff together.  I sometimes have two editors open, with different sets of related work in each.  The same way I used to use the Eclipse IDE - multiple projects from different repositories and locations in the filesystem, all showing in the one workspace.

**Comment by brainchild0** [comment]
Yes, I understand your remarks, but it is difficult to reconcile them with how you  suggest  Mark Text might behave differently from how it does now, compared to how Atom and Eclipse behave.

In the case of Eclipse, the workspace persists a listing of directories or files that are represented graphically whenever that workspace is restored. No metaphor for an Eclipse workspace exists in Mark Text. Are you asking for such a feature to be added?


 In the case of Atom, two separate menu items may open a directory, one in a new window, and one adding the directory tree to the sidebar of the current window.  Are you asking for a new menu item to be added to Mark Text, or an existing one to invoke a different behavior?

Note that Atom and Eclipse are principally designed for maintaining software projects, so expecting a one-to-one comparison in planned or supported features is not  reasonable generally, though specific features that support the objectives of any application certainly may be considered for inclusion.

**Comment by hmcts-jmoore** [comment]
To be honest, I'm not being all that prescriptive about the detail.  I work with markdown files which are part of software projects (mostly reading, not writing), and as notes across a variety of different activities spread across my hard disk.  I am likely to be switching between 'things' frequently through the day, and each 'thing' will typically involve bringing together content from different places - a few related git repositories with md documentation, some of my own notes, etc.  The tools that I work with all allow me to do that by having multiple windows open, each with a grouping of related folders.  Atom and Eclipse both have ways of remembering the groups of things I've had open together in the past - different ways, with Eclipse's workspaces being a much more heavyweight thing that Atom's projects.

In terms of implementation, I don't mind very much - Eclipse has a formal workspace that you have to explicitly create and add things to, while Atom just seems to remember and gives you the option (Reopen Project menu) to open a set of folders you've had open before.  They both work for me - the way I'm trying to use MarkText is closer to how I use Atom than Eclipse.

I entirely understand that I may be unusual in how I'm working with MarkText, so this may not be worth anyone looking into.  But it might also be a feature that a lot of people would find useful.

**Comment by brainchild0** [comment]
> To be honest, I'm not being all that prescriptive about the detail.

I understand.  Thinking about how your use  case might integrate with others in this application might help to resolve how to add some feature you find useful. Perhaps use of multiple windows and the recent-items list is sufficient. Perhaps a bookmarks feature would  be suitable, allowing you save some set of directories for rapid access.  Or perhaps you would think about some plugin that would offer a specific set of functions for a select group of users with needs similar to yours.

**Comment by geckolinux** [comment]
I agree that the simplest method would probably be to just allow multiple folders to opened in the sidebar. It would also be nice to allow multiple folders to be opened by default at startup.

**Comment by brainchild0** [comment]
I think it would be helpful  to reference an example  of an editor, preferably one targeted at general writing rather than writing software, that has behavior similar  to that you would wish to see in  see Mark Text.

**Comment by geckolinux** [comment]
@brainchild0 Thanks for the reply, the closest thing I can think of is QOwnNotes, which has a working folder switcher.

**Comment by brainchild0** [comment]
@geckolinux : Qownnotes has a feature called "Notes folders". Is your suggestion to include a similar feature in Mark Text?

**Comment by geckolinux** [comment]
@brainchild0  That's approximately the idea, yes.

**Comment by brainchild0** [comment]
@geckolinux: It seems that   the  model for managing files employed in QOwnNotes is different from most applications, in that it lacks a feature for opening an arbitrary file from the system file tree. It seems unlikely that this model would be adopted into Mark Text without being integrated with the normal feature  of simply opening a file. Perhaps it is best to consider how you think this integration might look best.

**Comment by geckolinux** [comment]
@brainchild0 Thanks for the reply. I agree that QOwnNotes isn't exactly an ideal solution for the way they handle multiple folders either.

So how about this?
1. When using "Open folder", simply make opened folders accumulate in the list instead of replacing the previously opened one.
2. Remember the previously opened folders when re-opening MarkText.
3. Use the existing **Save All** / **Close All** mouseover UI, but make it usable on all the folders. Here's a quick mockup:
     - ![Screenshot_2021-02-11_15-44-02](https://user-images.githubusercontent.com/2692138/107697405-2eff5b00-6c81-11eb-926d-d8ceb5b7b627.png)

**Comment by brainchild0** [comment]
@geckolinux: I wonder whether your idea is the best to reduce unwanted effects for the broadest range of users. Perhaps integration of the *recents* feature into the side bar offers a better separation of  functions while still providing largely the same functionality as you request. I am doubtful that users broadly will find the accumulation of folders into the same view as the open folder to be agreeable.

**Comment by geckolinux** [comment]
@brainchild0  That's not a bad idea either. But I think it would be necessary to distinguish between recent _files_ and recent _directories_, because when working with lots of different `.md` files they would make it harder to quickly find the top-level directories that have been used.

**Comment by brainchild0** [comment]
@geckolinux: Yes.  Such a distinction already occurs as the *recents* feature appears in the drop-down menu.

**Comment by geckolinux** [comment]
Currently I'm seeing recently opened `.md` files as well as recent folders in the **Open Recent >** menu.

![Screenshot_2021-02-11_18-10-50](https://user-images.githubusercontent.com/2692138/107711317-6bd54d00-6c95-11eb-8b7d-ed58c1c9010a.png)

**Comment by brainchild0** [comment]
@geckolinux: Sorry, it appears that my memory was inaccurate. Visual separation for files and folders would be a useful further enhancement.

**Comment by Redsandro** [comment]
@geckolinux commented:

> So how about this?
> 
> 1. When using "Open folder", simply make opened folders accumulate in the panel instead of replacing the previously opened one.
> 2. Remember the previously opened folders when re-opening MarkText, and also remember their expanded/closed state.
> 3. Use the existing **Save All** / **Close All** mouseover UI, but make it usable on all the folders. Here's a quick mockup:
>    
>    * ![Screenshot_2021-02-11_15-44-02](https://user-images.githubusercontent.com/2692138/107697405-2eff5b00-6c81-11eb-926d-d8ceb5b7b627.png)

Yes, this! :+1:  With the one expandable folder in the list on the left, I was pulling my hair out for not figuring out how to add another folder there. Otherwise why name the parent folder if there can only be one?

My intuition (or rather expectation) is partially the fault of Boostnote, where you can add multiple folders/projects to the sidebar:

![image](https://user-images.githubusercontent.com/1702193/117590764-31167000-b131-11eb-997a-9029b360379a.png)

**Comment by Redsandro** [comment]
> Under the proposed idea, would I be able to maintain different sets of open folders in different windows? Which configurations would be restored for a new application session?

I second @geckolinux notion that the same loaded folders should be visible on all Windows. In my opinion, simplicity should be key, unnecessary complexity should be avoided. One set of folders. Restored on start.

**Comment by geckolinux** [comment]
@brainchild0 If it's easy to implement a different list of loaded folders for different windows in the current session that would be fine too, but I assumed it would be a pain to code that functionality. Either way would be fine with me.

**Comment by brainchild0** [comment]
Yes, but in the case when each window would have  a different set of open directories,   an empty window for a new application session probably  would  need to be reset to some configured list of folders, rather than to one restored from the previous session.

The latter feature may be more suitable as part of a session manager, which would attempt to restore all the windows from a previous session at the start of a new one.

**Comment by Redsandro** [comment]
I am not _against_ such functionality either, as long as I don't get annoying things like:

1. Have 25 folders open in main window/session
2. Open new window with 1 folder/project
3. Restart computer
4. Open Mark Text and discover that I only have 1 folder open
5. All 25 carefully located folders are forgotten

And make no mistake, such things are _prone to happen_ when coding the ability for separate window sessions. Even Google's Chrome had this problem for _years_. To this day, I'm still afraid to open a second browser Window.

**However,** I _recommend_ against it. Apart from the added complexity, we must also think about a use case that causes the least confusion.

I think the majority of users have _one_ notes directory, as evidenced by the app not supporting even adding a second directory. Possibly synced with their favorite cloud option. How many dozens of temporary projects does an average user plan to add? And how confusing is it really to have folders like this in both windows:

- Private Notes
- Temp Project 1
- Temp Project 2

...and have the Private Notes expanded in one window, and have Temp Project 2 expanded in the other Window?

It has the added benefit of having "Temp Project x" in there for a month, until you're done working on it, at which point you can remove it again.

**Comment by Redsandro** [comment]
> It has the added benefit of having "Temp Project x" in there for a month so you won't have to relocate it, until you're done working on it, at which point you can remove it again.

We do need a right-click menu on the folder in the tree with the option to "unload" (AKA remove from the app).

**Comment by brainchild0** [comment]
>     All 25 carefully located folders are forgotten

Restoring state information created during a previous session is the job of a session manager. I believe such functionality should be considered separately.


See #988.

**Comment by Redsandro** [comment]
I'm rather new here so I don't know about the average decision making and development speed, but let's say I prefer something simpler in 3 months over something more complex in 18 months. Small steps. Release often. Amend and improve.

In the grand scheme of things, any multi-folder solution is an improvement over the current situation, and for that reason I agree with everything. :+1:

**Comment by Redsandro** [comment]
> it could happen, especially if someone submits a contribution.

Now that you mention it, I'm slightly disentranced to learn that there are 20 PRs pending since late last year. Even if I can figure out how to fix the [symlink thing](https://github.com/marktext/marktext/issues/2575) and use that as a workaround, the PR might not get merged in a while. :sweat_smile:

**Comment by brainchild0** [comment]
I'm sure any contribution would be  welcome, particularly one fixing an open issue. Having a stack of PRs to review is a much easier position for a maintainer than having a stack of issues with no proposed merges.

**Comment by Redsandro** [comment]
Since I am in the process of moving to a different note taking app and haven't settled on something yet, I decided that it's not worth waiting for a feature that I deem relatively crucial for my use case when development seems to be stalled and there are some serious alternatives.

Mark Text is definitely the prettiest so that's hard to replace. But I'm moving to [Zettlr](https://www.zettlr.com/). It has exactly what this issue describes: The ability to add multiple directories (called Workspaces).

**Comment by geckolinux** [comment]
> But I'm moving to Zettlr. It has exactly what this issue describes: The ability to add multiple directories (called Workspaces).

Interesting option, but unfortunately it doesn't meet my needs:
- Doesn't allow for creating ordered lists directly below another line as with many other versions of Markdown.
- Tab for indented list items works, but there's no way to back to the un-indented level without manually clicking and deleting, and strangely they are spaces and not tabs.
- Export to PDF required LaTex, which is far far far too bloated for me and my system.

**Comment by NeeraYu** [comment]
I am using Marktext to manage my notes written in markdown language. My notes consists of multiple folders scattered across different paths. Basically because some of the folders are shared with others and they cannot be put in a same parent directory due to the file sharing methods I am using.
So I think it will be better if Marktext could open multiple folders in one window. In this way I could do the global search & replace in this "workspace" as I can do in VSCode.
The main reason I want to stick to Marktext is WYSIWYG feature.

**Comment by mangoepic** [comment]
Might also be nice for managing notes on a cloud drive and on pc on the same time

**Comment by brainchild0** [comment]
I scanned have through the various comments, and I feel that some of the ideas support the use for multiple folders open at the same time in the *application*, but not as strongly in the same *window*. If various folders have separate meanings or categories of content, then a separate open window for each folder be adequate.

**Comment by samwh1te** [comment]
Multiple folders in one window would be amazing! The biggest thing missing for me from marktext right now
